### PR TITLE
Better explain how to run tests

### PR DESCRIPTION
**Pull request by olivierlacan**
It's a shame that it's so hard to find out how to run tests without running the entire test suite.

In a project like this, when making a contribution, it's often unnecessary to run the entire test suite when the 
CI will run it regardless on every commit, it's much more helpful to new contributors to explain how individual test
files they may be affecting can be run, and even individual test methods, since it's not discoverable through any of the 
feedback from Rake when running `rake test`.

For context this is the command line feedback I got when fumbling around with some Rake `TESTOPTS` flags I was trying: 

```
$ rake test TESTOPTS="--location test/rubygems/test_gem_ext_ext_conf_builder.rb"
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:121: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:134: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:149: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/templatable.rb:31: warning: assigned but unused variable - templatefile
missing argument: -n
Test::Unit automatic runner.
Usage: /Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/rake-13.0.3/lib/rake/rake_test_loader.rb [options] [-- untouched arguments]
    -r, --runner=RUNNER              Use the given RUNNER.
                                     (c[onsole], e[macs], x[ml])
        --collector=COLLECTOR        Use the given COLLECTOR.
                                     (de[scendant], di[r], l[oad], o[bject]_space)
    -n, --name=NAME                  Runs tests matching NAME.
                                     Use '/PATTERN/' for NAME to use regular expression.
                                     Regular expression accepts options.
                                     Example: '/taRget/i' matches 'target' and 'TARGET'
        --ignore-name=NAME           Ignores tests matching NAME.
                                     Use '/PATTERN/' for NAME to use regular expression.
                                     Regular expression accepts options.
                                     Example: '/taRget/i' matches 'target' and 'TARGET'
    -t, --testcase=TESTCASE          Runs tests in TestCases matching TESTCASE.
                                     Use '/PATTERN/' for TESTCASE to use regular expression.
                                     Regular expression accepts options.
                                     Example: '/taRget/i' matches 'target' and 'TARGET'
        --ignore-testcase=TESTCASE   Ignores tests in TestCases matching TESTCASE.
                                     Use '/PATTERN/' for TESTCASE to use regular expression.
                                     Regular expression accepts options.
                                     Example: '/taRget/i' matches 'target' and 'TARGET'
        --location=LOCATION          Runs tests that defined in LOCATION.
                                     LOCATION is one of PATH:LINE, PATH or LINE.
        --attribute=EXPRESSION       Runs tests that matches EXPRESSION.
                                     EXPRESSION is evaluated as Ruby's expression.
                                     Test attribute name can be used with no receiver in EXPRESSION.
                                     EXPRESSION examples:
                                       !slow
                                       tag == 'important' and !slow
        --[no-]priority-mode         Runs some tests based on their priority.
        --default-priority=PRIORITY  Uses PRIORITY as default priority
                                     (h[igh], i[mportant], l[ow], m[ust], ne[ver], no[rmal])
    -I, --load-path=DIR[:DIR...]     Appends directory list to $LOAD_PATH.
        --color-scheme=SCHEME        Use SCHEME as color scheme.
                                     (d[efault])
        --config=FILE                Use YAML format FILE content as configuration file.
        --order=ORDER                Run tests in a test case in ORDER order.
                                     (a[lphabetic], d[efined], r[andom])
        --max-diff-target-string-size=SIZE
                                     Shows diff if both expected result string size and actual result string size are less than or equal SIZE in bytes.
                                     (1000)
        --[no-]stop-on-failure       Stops immediately on the first non success test
                                     (false)
    -v, --verbose=[LEVEL]            Set the output level (default is verbose).
                                     (important-only, n[ormal], p[rogress], s[ilent], v[erbose])
        --[no-]use-color=[auto]      Uses color output
                                     (default is auto)
        --progress-row-max=MAX       Uses MAX as max terminal width for progress mark
                                     (default is auto)
        --no-show-detail-immediately Shows not passed test details immediately.
                                     (default is yes)
        --[no-]reverse-output        Shows fault details in reverse.
                                     (default is yes for tty output, no otherwise)
        --output-file-descriptor=FD  Outputs to file descriptor FD
        --                           Stop processing options so that the
                                     remaining options will be passed to the
                                     test.
    -h, --help                       Display this help.

Deprecated options:
        --console                    Console runner (use --runner).

Coverage report generated for Unit Tests to /Users/olivierlacan/Development/oss/rubygems_org/rubygems/coverage. 2168 / 9070 LOC (23.9%) covered.
SimpleCov failed with exit 1rake aborted!
Command failed with status (1)

Tasks: TOP => test
(See full trace by running task with --trace)
```

The output is extremely verbose and none of the flag examples shown work: 
```
$ rake test TESTOPTS="-n test_class_build_extconf_fail"
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:121: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:134: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/generator.rb:149: warning: assigned but unused variable - rel_prefix
/Users/olivierlacan/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/sdoc-2.1.0/lib/sdoc/templatable.rb:31: warning: assigned but unused variable - templatefile

File does not exist: /Users/olivierlacan/Development/oss/rubygems_org/rubygems/test_class_build_extconf_fail

Coverage report generated for Unit Tests to /Users/olivierlacan/Development/oss/rubygems_org/rubygems/coverage. 2168 / 9070 LOC (23.9%) covered.
SimpleCov failed with exit 1rake aborted!
Command failed with status (1)

Tasks: TOP => test
(See full trace by running task with --trace)
```

I can surmise this is because of how `rake_test_loader` works and how the [Rakefile is set up](https://github.com/olivierlacan/rubygems/blob/fe17b504fe96653e3b482aaad68d0cd1f496064e/Rakefile#L33-L41). But that's not at all obvious or easy to figure out, and despite having used Ruby codebases for years, it took me quite a while to figure out.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -31,14 +30,27 @@ here: https://guides.rubygems.org/contributing/
 ## Getting Started
 
     $ rake setup
-    $ rake test
 
-> Optional you can configure git hooks with: rake git_hooks
+> Optionally you can configure git hooks with: rake git_hooks
```
Comment (olivierlacan) [inline CONTRIBUTING.md:34]: Thought I might as well fix this typo while I was in there.

**Review by deivid-rodriguez — COMMENTED** [review]
This looks good to me, thanks for documenting this!

One question is, did you try this against the latest commit in the default branch? Only a few days ago we migrated rubygems test suite to use `test-unit` instead of `minitest` under the hood, so I'm unsure whether any of this might've changed.

**Comment by deivid-rodriguez** [comment]
Sorry, forget about that, I just read the output you got and it indeed points out that `test-unit` is being used, so it seems all fine :+1:. I'm glad the way to run specific tests hasn't changed :)

**Comment by deivid-rodriguez** [comment]
I have no idea why status checks are not getting updated :thinking:.

**Comment by simi** [comment]
:information_source: for testing I do use

```ruby
rake setup # install deps
rake test # test whole suite
ruby -Ilib:test test/rubygems/test_deprecate.rb # test selected file(s)
ruby -Ilib:test test/rubygems/test_deprecate.rb -n /test_default/ # test selected file(s) and pick only selected method(s)
```

**Comment by deivid-rodriguez** [comment]
I used @olivierlacan's way until now, but @simi's approach feels nicer, actually.

**Comment by simi** [comment]
@deivid-rodriguez there is one trick you can use using this approach:

```ruby
ruby -Ilib:test test/rubygems/test_deprecate.rb # test selected file(s) against repo RubyGems
ruby -Itest test/rubygems/test_deprecate.rb # test selected file(s) against currently installed RubyGems
```

You can check if test is failing with installed RG as well. Anyway I'm not sure how safe that is :)

**Comment by olivierlacan** [comment]
@deivid-rodriguez I used [ci skip] in the commit body. I can remove it if you want CI to run?

**Comment by olivierlacan** [comment]
@simi There we go. I knew we’d get some cool tricks. I’ll add that. 🤗

**Comment by deivid-rodriguez** [comment]
Oh, I see. Yeah, please remove it :+1:.

**Comment by olivierlacan** [comment]
@deivid-rodriguez Force pushed the removal, looks like CI is running now. Thanks for the tip. 😃

**Comment by olivierlacan** [comment]
@simi Added your direct `ruby` runner examples. I'm seeing a few unrelated test failures in rubygems/core. I'll leave it up to @deivid-rodriguez I guess?

**Comment by deivid-rodriguez** [comment]
Yeah, that's currently broken at the moment :(. I'll be ignoring that failure until we can address it.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -31,14 +30,31 @@ here: https://guides.rubygems.org/contributing/
...
+
+To run an individual test file located for example in `test/rubygems/test_deprecate.rb` you can use: 
+
+    $ ruby -Ilib:test test/rubygems/test_deprecate.rb 
+    
+And to run an individual test method named `test_default` within a test file, you can use: 
+
+    $ ruby -Ilib:test test/rubygems/test_deprecate.rb -n /test_default/ 
+
+You can even run a test against your currently installed RubyGems version instead of the repository itself if you want to check whether the test would be failing on it as well:
```
Comment (deivid-rodriguez) [inline CONTRIBUTING.md:54]: I'm not sure how safe this is. I'd say it shouldn't perform any destructive actions on your system install, but I'm not fully sure. Maybe keep that trick "a secret" at least for now?

Comment (olivierlacan) [reply CONTRIBUTING.md]: Ok, removed. :-)

**Review by deivid-rodriguez — APPROVED** [review]

**Review by simi — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Thank you :heart:.

**Pull request merged**